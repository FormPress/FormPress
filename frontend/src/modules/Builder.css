.content > .builder {
  margin-top: 86px;
}

.builder > .headerContainer,
.forms > .headerContainer,
.data > .headerContainer {
  height: 30px;
  top: 86px;
  width: 100%;
  position: fixed;
  background-color: #e2e2e2;
  z-index: 1;
}

.builder > .headerContainer > .header,
.data > .headerContainer > .header {
  height: 30px;
  width: 1200px;
}

.builder > .headerContainer > .header .back,
.data > .headerContainer > .header .back {
  background-color: #c9c9c9;
  color: rgb(48, 93, 134);
  line-height: 30px;
  height: 30px;
  width: 57px;
  text-align: center;
  cursor: pointer;
  display: inline-block;
}

.builder .header > .mainTabs > a,
.data .header > .mainTabs > a {
  color: #1c5c85;
  display: inline-block;
  height: 30px;
  line-height: 30px;
  padding: 0 10px;
  text-decoration: none;
}

.builder .header > .mainTabs > a.selected,
.data .header > .mainTabs > a.selected {
  background-color: #1c5c85;
  color: #ffffff;
}

.header > .formTitle {
  width: 400px;
  min-height: 20px;
}

.header > .formControls {
  width: 400px;
  margin-left: 20px;
}

.builder > .content {
  padding-top: 40px;
  width: 1200px;
  margin: 0 auto;
}

/* Builder */

.builder > .content > .leftTabs {
  min-height: 100px;
}

.builder > .content > .leftTabs > div {
  list-style: none;
  box-shadow: 0px 3px 10px #00000014;
  background-color: white;
  margin: 0;
  padding: 0;
  position: fixed;
  width: 57px;
}

.builder > .content > .leftTabs > div > a {
  width: 59px;
  height: 57px;
  font-size: 25px;
  line-height: 57px;
  text-align: center;
  display: inline-block;
}

.builder > .content > .leftTabs > div > a,
.builder > .content > .leftTabs > div > a:visited {
  color: rgb(148, 200, 50);
}
.builder > .content > .leftTabs > div > a:hover {
  color: white;
  background-color: rgb(148, 200, 50);
}

.builder > .content > .leftTabs > div > a.selected {
  color: white;
  background-color: rgb(168, 227, 57);
}

.builder > .content > .leftMenu {
  min-height: 500px;
}

.builder > .content > .leftMenu > .leftMenuContents {
  position: fixed;
  top: 126px;
  width: 359px;
  box-shadow: 0px 3px 10px #00000014;
}

.elements > .elementsMessage {
  background-color: #fffcdd;
  color: #646464;
  font-size: 14px;
  text-align: center;
  padding: 15px 20px;
}

.elements > .elementList {
  margin-top: 20px;
}

.elements > .elementList > .element {
  margin: 10px 12px 10px 16px;
  padding: 10px 10px 10px 12px;
  border-radius: 8px;
  font-family: 'Montserrat';
  border: solid 1px #ffffff;
  box-shadow: 0 3px 6px 0 rgba(3, 39, 57, 0.04);
  background-color: #ffffff;
  color: #646464;
  width: 91%;
  cursor: move;
}

.elements > .elementList > .element:hover {
  border: solid 1px #005a87;
  color: #1c5c85;
}

.elements > .elementList > .element:hover .add-element-button {
  visibility: visible;
}

.element-picker-icon-wrapper {
  width: 12%;
  display: inline-block;
  text-align: center;
}

.element-picker-icon {
  color: #7a9eba;
}

.element-picker-text {
  margin: 2px 0 3px 12px;
  display: inline-block;
  width: 76%;
  font-size: 16px;
  font-weight: 500;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.38;
  letter-spacing: normal;
  text-align: left;
}
.add-element-button {
  height: 16px;
  margin: 4px 0 4px 2px;
  color: #c9c9c9;
  visibility: hidden;
  cursor: pointer;
}

.add-element-button:hover {
  color: #68942f;
}

.builderStage > .formTitle {
  height: 60px;
  line-height: 60px;
  background-color: #fdfdfd;
  color: #1c5c85;
  padding-left: 16px;
  font-size: 18px;
}

.builderStage > .formControls {
  margin-top: 10px;
  background-color: white;
  text-align: right;
  height: 32px;
  padding-right: 0;
}

.builderStage > .formControls > button {
  border: none;
  height: 32px;
  line-height: 32px;
  font-size: 14px;
  font-weight: bold;
  background-color: white;
  cursor: pointer;
  color: #8cc63f;
  padding: 0 20px;
  margin-left: 10px;
}

.builderStage > .formControls > button.publish {
  background-color: #8cc63f;
  color: #ffffff;
  position: relative;
}

.builderStage > .formControls > button.publish > .publishRequired {
  background-color: red;
  width: 10px;
  height: 10px;
  top: 11px;
  right: 5px;
  float: right;
  border-radius: 10px;
  position: absolute;
}

.builderStage > .form {
  box-shadow: 0px 3px 10px #00000014;
  margin: 0 auto;
}

.form.builder.dragging > * * {
  pointer-events: none;
}

.form.builder > .dropPlaceHolder {
  pointer-events: none;
  height: 80px;
  width: 100%;
  background-color: blue;
  opacity: 0.1;
  border: 3px solid blue;
}

.form.builder > .element {
  border: 1px solid transparent;
}

.form.builder > .element .action {
  display: none;
  border: solid 1px #e2e2e2;
  background-color: #f5f5f5;
  border-radius: 12px;
  font-size: 14px;
  position: absolute;
  right: 10px;
  top: 15px;
  padding: 0 5px;
}

.form.builder > .element .fl.input .action {
  right: 0px;
  top: -2px;
  display: none;
}

.form.builder > .element .fl.input li:hover .action,
.form.builder > .element .fl.input:hover .checkbox-label + .action {
  display: block;
}

.form.builder .element .action > * {
  cursor: pointer;
  margin: 5px;
  font-size: 14px;
}

.form.builder > .element > .action > .sortHandle {
  cursor: move;
  display: inline-flex;
}

.form.builder > .element > .action > .moveButton {
  display: inline-block;
  font-size: 16px;
}

.form.builder > .element:hover,
.form > .element.selected {
  border: 1px solid black;
}

.form.builder > .element:hover > .action,
.form > .element.selected > .action {
  display: inline-flex;
  float: right;
}

.popoverText {
  display: none;
}

.popover-container {
  display: inline-flex;
  position: relative;
}

.popover-container:hover > .popoverText {
  display: block;
  width: 83px;
  height: 20px;
  margin: 0 0 3px;
  padding: 2px 4px 3px 3px;
  border-radius: 4px;
  position: absolute;
  background-color: #113952;
  color: #fff;
  text-align: center;
  top: -25px;
  left: -35px;
  z-index: 9999;
}

.leftMenuContents .element {
  padding: 20px 0px 10px 10px;
}
#qc_required .input {
  float: right;
}
