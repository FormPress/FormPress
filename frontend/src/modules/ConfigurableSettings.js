import datasets from '../datasets'

const policy = {
  required: {
    rule: {
      type: 'exceptAll',
      exceptions: ['Button', 'Header', 'Separator']
    },
    configurableSettings: {
      default: false,
      formProps: {
        type: 'Checkbox',
        label: '',
        options: ['Make this field required']
      }
    }
  },
  requiredText: {
    rule: {
      type: 'exceptAll',
      exceptions: ['Button', 'Header', 'Separator']
    },
    configurableSettings: {
      default: 'Required field',
      formProps: {
        type: 'TextBox',
        label: 'Error message when this field is left empty.',
        placeholder: 'Required field'
      }
    }
  },
  placeholder: {
    rule: {
      type: 'only',
      exceptions: ['TextBox', 'TextArea', 'Email', 'Dropdown']
    },
    configurableSettings: {
      default: '',
      formProps: {
        type: 'TextBox',
        label: 'Placeholder Text',
        placeholder: 'Enter a placeholder text'
      }
    }
  },
  hasDataset: {
    rule: {
      type: 'only',
      exceptions: ['Dropdown']
    },
    configurableSettings: {
      default: false,
      formProps: {
        type: 'Checkbox',
        label: '',
        options: ['Predefined options']
      }
    }
  },
  dataset: {
    rule: {
      type: 'only',
      exceptions: ['Dropdown']
    },
    configurableSettings: {
      default: 'countries',
      formProps: {
        type: 'Dropdown',
        label: 'Choose a data set',
        placeholder: false,
        options: datasets
      }
    }
  }
}

export const getConfigurableSettings = (questionType) => {
  let willReturnObject = {}

  Object.keys(policy).map(function (objectKey) {
    const rule = policy[objectKey].rule
    if (
      rule.type === 'exceptAll' &&
      rule.exceptions.indexOf(questionType) === -1
    ) {
      willReturnObject[objectKey] = policy[objectKey].configurableSettings
    } else if (
      rule.type === 'only' &&
      rule.exceptions.indexOf(questionType) > -1
    ) {
      willReturnObject[objectKey] = policy[objectKey].configurableSettings
    } else if (rule.type === 'all') {
      willReturnObject[objectKey] = policy[objectKey].configurableSettings
    }
    return {}
  })

  return willReturnObject
}
