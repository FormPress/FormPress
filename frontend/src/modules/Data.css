.data {
  width: 100%;
}

.data .header {
  margin: 0 auto;
}

.data > .formSelectorContainer {
  height: 60px;
  top: 132px;
  width: 100%;
  position: fixed;
}

.data > .formSelectorContainer > .formSelector {
  border-left: 8px solid #113952;
  background-color: white;
  color: #333333;
  font-size: 18px;
  padding-left: 8px;
  height: 60px;
  line-height: 60px;
  cursor: pointer;
}

.data > .formSelectorContainer > .formSelector > .down {
  background-color: #719fbd;
  color: white;
  text-align: center;
  font-size: 22px;
}

.data > .formSelectorContainer > .formSelectorOptions {
  background-color: white;
  border-left: 8px solid #e2e2e2;
  border-bottom: 8px solid #719fbd;
  max-height: 600px;
  overflow: scroll;
}

.data > .formSelectorContainer > .formSelectorOptions ul {
  margin: 0;
  padding: 0;
}

.data > .formSelectorContainer > .formSelectorOptions li {
  list-style: none;
  border: 1px solid transparent;
  border-top: 1px solid rgb(234, 234, 234);
  padding-left: 10px;
  height: 60px;
  line-height: 60px;
  cursor: pointer;
}

.data > .formSelectorContainer > .formSelectorOptions li:hover {
  border: 1px solid black;
}

.data > .dataContent {
  margin-top: 210px;
  background-color: white;
}

.data > .dataContent table.fp_table > tbody > tr {
  height: 32px;
  background-color: #ffffff;
  cursor: pointer;
  font-size: 14px;
}

.data > .dataContent table.fp_table > tbody > tr.selected {
  background-color: #ddd;
  color: #113952;
}

.data > .dataContent table.fp_table > tbody > tr > td > span.table_caret_right {
  display: table;
  margin: auto;
  font-size: 18px;
}

.data > .dataContent table.fp_table > tbody > tr {
  color: #a0a0a0;
}

.data > .dataContent > .submissionSelector > .submissionActions > * {
  padding: 16px;
  color: #646464;
  font-size: 10px;
}

.data
  > .dataContent
  > .submissionSelector
  > .submissionActions
  > .buttonContainer {
  text-align: right;
  padding-right: 0;
}

.data
  > .dataContent
  > .submissionSelector
  > .submissionActions
  .csvExportButton {
  border: none;
  background-color: #8cc63f;
  color: white;
  font-size: 10px;
  font-weight: bold;
  text-align: center;
  border-radius: 10px;
  padding: 8px 12px;
  cursor: pointer;
}

.data
  > .dataContent
  > .submissionSelector
  > .submissionActions
  .csvExportButton.disabled {
  cursor: not-allowed;
  background-color: #ccc;
}

.data > .dataContent > .entriesViewer,
.data > .dataContent > .submissionSelector > .noData {
  padding-top: 20px;
  padding-bottom: 20px;
  padding-left: 18px;
}

.data > .dataContent > .entriesViewer > .entry {
  margin-bottom: 20px;
  border-left: 2px solid #e2e2e2;
  text-indent: 12px;
  font-family: Montserrat;
  font-size: 12px;
  font-weight: 600;
  line-height: 1.83;
}

.data > .dataContent > .entriesViewer > .entry .value input[type='checkbox'] {
  width: 18px;
  height: 18px;
  -webkit-appearance: none;
  -o-appearance: none;
  appearance: none;
  border: 1px solid #a0a0a0;
  border-radius: 4px;
  outline: none;
  transition-duration: 0.3s;
  background-color: #fdfdfd;
  cursor: pointer;
  overflow: hidden;
  margin-top: 2px;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox']:checked::before {
  content: '\2713';
  display: block;
  text-align: center;
  color: #ffffff;
  font-size: 16px;
  margin: auto;
  position: relative;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox']:checked {
  border: 1px solid #a0a0a0;
  background-color: #8cc63f;
}

.data > .dataContent > .entriesViewer > .entry .value .checkbox-label {
  color: #333333;
  display: inline-block;
  margin-top: 0px;
  position: absolute;
  text-indent: 5px;
  white-space: nowrap;
  max-width: 600px;
  overflow: hidden;
  user-select: none;
  min-width: 100px;
  border: 1px solid #ffffff;
}

.data > .dataContent > .entriesViewer > .entry .value input[type='radio'] {
  position: relative;
  width: 18px;
  height: 18px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -o-appearance: none;
  appearance: none;
  border: 1px solid #a0a0a0;
  border-radius: 50%;
  transition-duration: 0.3s;
  outline: none;
  background-color: #fdfdfd;
  cursor: pointer;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='radio']:checked::after {
  content: ' ';
  width: 10px;
  height: 10px;
  border: 3px solid #ffffff;
  border-radius: 50%;
  background-color: #8cc63f;
  position: absolute;
  margin: 3px;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='radio']:checked {
  background-color: #8cc63f;
  border: 1px solid #8cc63f;
}

.data > .dataContent > .entriesViewer > .entry .value .radio-label {
  color: #333333;
  margin-top: -2px;
  position: absolute;
  -webkit-user-select: none;
  user-select: none;
  border: 1px solid #ffffff;
  min-width: 100px;
  text-indent: 5px;
}

.data > .dataContent > .entriesViewer > .entry > .label {
  color: #a0a0a0;
}
.data > .dataContent > .entriesViewer > .entry > .value {
  color: #113952;
  margin-left: 12px;
  text-indent: 0px;
}

.data .text_center input[type='checkbox'] {
  width: 18px;
  height: 18px;
  -webkit-appearance: none;
  -o-appearance: none;
  appearance: none;
  border: 1px solid #a0a0a0;
  border-radius: 4px;
  outline: none;
  transition-duration: 0.3s;
  background-color: #fdfdfd;
  cursor: pointer;
  overflow: hidden;
  margin-top: 2px;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox'].toggle-checkbox {
  opacity: 0;
  width: 40px;
  height: 18px;
  z-index: 9999;
  position: relative;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox'].toggle-checkbox:checked
  + .slider {
  background-color: #cae1ab;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox'].toggle-checkbox
  + .slider {
  position: absolute;
  cursor: pointer;
  background-color: #d9dedc;
  transition: 0.4s;
  width: 40px;
  height: 18px;
  border-radius: 16px;
  margin-top: -24px;
  display: block;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox'].toggle-checkbox:checked
  + .slider:before {
  transform: translateX(18px);
  background-color: #8cc63f;
}

.data > .dataContent > .entriesViewer > .entry .value .toggle-label {
  color: #333333;
  display: inline-block;
  position: absolute;
  text-indent: 5px;
  white-space: nowrap;
}

.data
  > .dataContent
  > .entriesViewer
  > .entry
  .value
  input[type='checkbox'].toggle-checkbox
  + .slider:before {
  position: absolute;
  content: '';
  height: 22px;
  width: 22px;
  left: 0px;
  bottom: -2px;
  background-color: #bec6c4;
  transition: 0.4s;
  border-radius: 12px;
}

.data .text_center input[type='checkbox']:checked {
  border: 1px solid #a0a0a0;
  background-color: #8cc63f;
}

.data .text_center input[type='checkbox']:checked::before {
  content: '\2713';
  display: block;
  text-align: center;
  color: #ffffff;
  font-size: 16px;
  margin: auto;
  position: relative;
}
