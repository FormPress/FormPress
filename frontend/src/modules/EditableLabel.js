import React, { Component } from 'react'

export default class EditableLabel extends Component {
  constructor (props) {
    super(props)

    this.handleOnInput = this.handleOnInput.bind(this)
    this.handleOnBlur = this.handleOnBlur.bind(this)
  }

  handleOnInput (e) {

  }

  handleOnBlur (e) {
    this.props.handleLabelChange(
      this.props.labelKey,
      e.target.innerHTML.replace(/(<([^>]+)>)/ig, '')
    )
  }

  render() {
    const extraProps = {}

    if (this.props.mode === 'builder') {
      extraProps.contentEditable = true  
    }

    return (
      <div
        className={ this.props.className }
        onInput={ this.handleOnInput }
        onBlur={ this.handleOnBlur }
        { ...extraProps }
      >
        { this.props.value }
      </div>
    )
  }
}
