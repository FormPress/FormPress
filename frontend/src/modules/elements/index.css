.form > .element {
  box-sizing: border-box;
  padding: 5px 10px;
}

.form > .element > .label {
  width: 200px;
}

.form > .element > .input {
  width: 400px;
}