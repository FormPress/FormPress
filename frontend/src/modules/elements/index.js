import './index.css'

export {default as Button} from './Button.js'
export {default as Checkbox} from './Checkbox.js'
export {default as Text} from './Text.js'
export {default as TextArea} from './TextArea.js'
export {default as Dropdown} from './Dropdown.js'
export {default as Radio} from './Radio.js'
export {default as Name} from './Name.js'
export {default as FileUpload} from './FileUpload.js'