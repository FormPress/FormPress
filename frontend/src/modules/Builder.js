import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons'

import * as Elements from './elements'
import AuthContext from '../auth.context'
import Renderer from './Renderer'
import EditableLabel from './common/EditableLabel'
import FormProperties from './helper/FormProperties'
import QuestionProperties from './helper/QuestionProperties'
import { api } from '../helper'

import './Builder.css'

const BACKEND = process.env.REACT_APP_BACKEND

const getElements = () => Object.values(Elements).map((element) => {
    const config = Object
      .assign({}, element.defaultConfig)
    
    if (typeof element.configurableSettings !== 'undefined') {
      const configurableKeys = Object.keys(element.configurableSettings)

      for (const key of configurableKeys) {
        config[key] = element.configurableSettings[key].default
      }
    }

    return config
  }
)
const getElementsConfigurableSettingsObject = () => Object
  .values(Elements)
  .reduce((acc, element) => {
    acc[element.defaultConfig.type] = {
      configurableSettings: element.configurableSettings || {}
    }

    return acc
  }, {})

const getWeightedElements = () => Object
  .values(Elements)
  .map((element) => Object
    .assign({}, element.defaultConfig, {weight: element.weight})
  )
const getElementsKeys = () => getElements()
  .reduce((acc, item) => {
    acc[item.type] = item

    return acc
  }, {})

//Stuff that we render in left hand side
const pickerElements = getWeightedElements()
  .sort((a, b) => a.weight - b.weight)

class Builder extends Component {
  async componentDidMount () {
    if (typeof this.props.match.params.formId !== 'undefined') {
      const { formId } = this.props.match.params

      if (formId !== 'new') {
        await this.loadForm(formId)
        window.localStorage.setItem('lastEditedFormId', formId)  
      } else {
        window.scrollTo(0, 0)
        this.setIntegration({ type: 'email', to: this.props.auth.email })
      }
    } else {
      const lastEditedFormId = window.localStorage.getItem('lastEditedFormId')

      if (lastEditedFormId !== null) {
        this.props.history.push(`/editor/${lastEditedFormId}`)
        setTimeout(() => {
          this.componentDidMount()
        }, 1)
      }
    }
  }

  async loadForm (formId) {
    this.setState({ loading: true })

    const { data } = await api({
      resource: `/api/users/${this.props.auth.user_id}/forms/${formId}`
    })

    if (typeof data.props === 'undefined') {
      this.setState({
        loading: false
      })
      return
    }

    const props = JSON.parse(data.props)
    const form = {
      ...data,
      props
    }

    this.setState({
      loading: false,
      form
    })
  }

  setIntegration (_integration) {
    const form = { ...this.state.form }

    form.props.integrations = [...form.props.integrations]

    const matched = form.props.integrations.filter((integration) =>
      (integration.type === _integration.type)
    )

    if (matched.length > 0) {
      const index = form.props.integrations.indexOf(matched[0])

      form.props.integrations[index] = {
        ...form.props.integrations[index],
        ..._integration
      }
    } else {
      form.props.integrations.push(_integration)
    }

    this.setState({ form })
  }

  setActiveTab (activeTab) {
    const newState = { activeTab }

    if (activeTab !== 'questionProperties') {
      newState.selectedFieldId = false
    }

    this.setState(newState)
  }

  constructor (props) {
    super(props)
    this.state = {
      counter: 0,
      activeTab: 'elements',
      saving: false,
      loading: false,
      dragging: false,
      dragIndex: false,
      insertBefore: false,
      selectedFieldId: false,
      form: {
        id: null,
        user_id: null,
        title: 'Untitled Form',
        props: {
          integrations: [
            {
              type: 'email',
              value: ''
            }
          ],
          elements: [
            {
              id: 1,
              type: 'Text',
              label: 'First Name'
            },
            {
              id: 2,
              type: 'Button',
              buttonText: 'Submit'
            }
          ]
        }
      }
    }
    //this.handleClick = this.handleClick.bind(this)
    this.handleDragStart = this.handleDragStart.bind(this)
    this.handleDragOver = this.handleDragOver.bind(this)
    this.handleDrop = this.handleDrop.bind(this)
    this.handleDragEnd = this.handleDragEnd.bind(this)
    this.handleSaveClick = this.handleSaveClick.bind(this)
    this.handlePreviewClick = this.handlePreviewClick.bind(this)
    this.handleLabelChange = this.handleLabelChange.bind(this)
    this.handleTitleChange = this.handleTitleChange.bind(this)
    this.handleFormElementClick = this.handleFormElementClick.bind(this)
    this.handleFormElementDeleteClick = this.handleFormElementDeleteClick.bind(this)
    this.setIntegration = this.setIntegration.bind(this)
    this.setActiveTab = this.setActiveTab.bind(this)
    this.configureQuestion = this.configureQuestion.bind(this)
  }

  handleDragStart (item, e) {
    e.dataTransfer.setData('text/plain', item.type)
    this.setState({ dragging: true })
  }

  handleDrop (e) {
    e.stopPropagation()
    e.preventDefault()
    
    const type = e.dataTransfer.getData('text')
    const item = getElementsKeys()[type]
    const { form, dragIndex } = this.state

    //set auto increment element id
    const maxId = Math.max(...form.props.elements.map((element) => element.id))

    item.id = maxId + 1

    const index = form.props.elements.findIndex(
      (element) => (element.id.toString() === dragIndex)
    )

    let newElements

    if (this.state.insertBefore === true) {
      newElements = [
        ...form.props.elements.slice(0, index),
        item,
        ...form.props.elements.slice(index)
      ]
    } else {
      newElements = [
        ...form.props.elements.slice(0, index + 1),
        item,
        ...form.props.elements.slice(index + 1)
      ]
    }

    this.setState({
      dragging: false,
      dragIndex: false,
      form: {
        ...form,
        props: {
          ...form.props,
          elements: newElements
        }
      }
    })
  }

  handleDragEnd (e) {
    this.setState({ dragging: false })
  }

  handleDragOver (e) {
    const rect = e.target.getBoundingClientRect()
    const {top, height} = rect
    const {clientY} = e
    const id = e.target.id.replace('qc_', '')
    const middleTop = top + height / 2
    const diff = clientY - middleTop
    const insertBefore = (diff < 0)

    if (
      id !== '' &&
      (
        id !== this.state.dragIndex.toString() ||
        insertBefore !== this.state.insertBefore
      )
    ) {
      this.setState({
        dragIndex: id,
        insertBefore
      })
    }

    e.stopPropagation()
    e.preventDefault()
  }

  handleLabelChange (id, value) {
    const form = { ...this.state.form }

    form.props.elements = [...form.props.elements]

    const question = form
      .props
      .elements
      .filter((element) => (element.id === id))[0]

    if (question.type === 'Button') {
      question.buttonText = value
    } else {
      question.label = value
    }

    this.setState({ form })
  }

  handleTitleChange (id, value) {
    const form = { ...this.state.form }

    form.title = value

    this.setState({ form })
  }

  handleFormElementClick (e) {
    e.preventDefault()
    const id = parseInt(e.target.id.replace('qc_', ''))
    const { elements } = this.state.form.props

    const matchingElements = elements.filter((elem) => (elem.id === id))

    if (matchingElements.length === 1) {
      this.setState({
        selectedFieldId: id,
        activeTab: 'questionProperties'
      })
    }
  }

  handleFormElementDeleteClick (id) {
    const form = { ...this.state.form }

    form.props.elements = form.props.elements.filter((elem) => (elem.id !== id))

    this.setState({ form })
  }

  async handleSaveClick (e) {
    const { form } = this.state

    this.setState({ saving: true })

    const { data } = await api({
      resource: `/api/users/${this.props.auth.user_id}/forms`,
      method: (form.id === null) ? 'post' : 'put',
      body: this.state.form
    })

    this.setState({ saving: false })

    if (form.id === null && typeof data.id !== 'undefined') {
      this.props.history.push(`/editor/${data.id}`)
      this.setState({
        form: {
          ...this.state.form,
          id: data.id
        }
      })
      window.localStorage.setItem('lastEditedFormId', data.id)
    }
  }

  handlePreviewClick (e) {
    const { id } = this.state.form

    window.open(`${BACKEND}/form/view/${id}`, '_blank')
  }

  configureQuestion (changes) {
    const form = { ...this.state.form }

    form.props.elements = [...form.props.elements]

    const question = form
      .props
      .elements
      .filter((element) => (element.id === changes.id))[0]

    Object.assign(question, changes.newState)

    this.setState({ form })
  }

  render () {
    const {
      activeTab,
      dragging,
      form,
      loading,
      saving,
      selectedFieldId
    } = this.state
    const saveButtonProps = {}
    const tabs = [
      { name: 'elements', text: 'Elements' },
      { name: 'formProperties', text: 'Form Properties' }
    ]

    if (selectedFieldId !== false) {
      tabs.push({ name: 'questionProperties', text: 'Question Properties' })
    }

    if (saving === true || loading === true) {
      saveButtonProps.disabled = true
    }

    return (
      <div className='builder'>
        <div className='headerContainer'>
          <div className='header grid center'>
            <div className='col-1-16'>
              <Link to='/forms' className='back'>
                <FontAwesomeIcon icon={ faChevronLeft } />
              </Link>
            </div>
            <div className='col-15-16 mainTabs'>
              {tabs.map((item, key) => (
                <a
                  href='#/'
                  key={ key }
                  onClick={ this.setActiveTab.bind(this, item.name) }
                  className={
                    (item.name === activeTab) ? 'selected' : undefined
                  }
                >
                  { item.text }
                </a>
              ))}
            </div>
          </div>
        </div>
        <div className='content grid'>
          <div className='leftTabs col-1-16'>
          </div>
          <div className='leftMenu col-5-16'>
            <div className='leftMenuContents'>
              { this.renderLeftMenuContents() }
            </div>
          </div>
          <div className='builderStage col-10-16 grid'>
            <div className='formTitle col-16-16'>
              {
                (loading === false)
                  ? <EditableLabel
                    className='label'
                    mode='builder'
                    labelKey='title'
                    handleLabelChange={ this.handleTitleChange }
                    value={ form.title }
                  />
                  : null
              }
            </div>
            <div className='col-16-16 formControls'>
              <button onClick={ this.handleSaveClick } { ...saveButtonProps }>
                { saving === true ? 'Saving...': 'Save' }
              </button>
              <button onClick={ this.handlePreviewClick }>
                Preview
              </button>
              <button className='publish'>
                Publish
              </button>
            </div>
            {
            (loading === true)
              ? 'Loading...'
              : <Renderer
                className={`col-16-16 form${(dragging === true)? ' dragging' : ''}`}
                builderHandlers={{
                  onDrop: this.handleDrop,
                  onDragOver: this.handleDragOver,
                  onClick: this.handleFormElementClick
                }}
                customBuilderHandlers={{
                  onDelete: this.handleFormElementDeleteClick
                }}
                handleLabelChange={ this.handleLabelChange }
                dragIndex={ this.state.dragIndex }
                dragging={ dragging }
                insertBefore={ this.state.insertBefore }
                form={ form }
                selectedFieldId={ selectedFieldId }
                mode='builder'
              />
          }
          </div>
        </div>
      </div>
    )
  }

  renderLeftMenuContents () {
    const { activeTab, form, selectedFieldId } = this.state
    const selectedField = {}

    if (selectedFieldId !== false) {
      const selectedFieldConfig = form
        .props
        .elements
        .filter((elem) => (elem.id === selectedFieldId))[0]

      const elements = getElementsConfigurableSettingsObject()

      selectedField.config = selectedFieldConfig
      selectedField.configurableSettings = elements[selectedFieldConfig.type]
        .configurableSettings || {}
    }
    
    switch (activeTab) {
      case 'elements':
        return (
          <div className='elements'>
            <div className='elementsMessage'>
              Drag and Drop elements to right hand side to add to the form.
              Or you can click + icon
            </div>
            <div className='elementList'>
              {pickerElements.map((elem) =>
                <div
                  className='element'
                  draggable
                  onDragStart={ this.handleDragStart.bind(this, elem) }
                  onDragEnd={ this.handleDragEnd }
                  key={ elem.type }
                >
                  { elem.type }
                </div>
              )}
            </div>
          </div>
        )
      case 'formProperties':
        return (
          <FormProperties
            form={ form }
            setIntegration={ this.setIntegration }
          />
        )
      case 'questionProperties':
        return (
          <QuestionProperties
            selectedField={ selectedField }
            configureQuestion={ this.configureQuestion }
          />
        )
    }
  }
}

const BuilderWrapped = (props) => 
  <AuthContext.Consumer>
    {
      (value) => <Builder { ...props } auth={ value } />
    }
  </AuthContext.Consumer>

export default BuilderWrapped
