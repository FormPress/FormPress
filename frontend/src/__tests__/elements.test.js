import React from 'react'
import { create } from 'react-test-renderer'
import * as Elements from '../modules/elements'

const excludedKeys = ["Button", "Header", "Separator"]
const keys = Object.keys(Elements)
for (const key of keys) {
  const Component = Elements[key]

  describe(`Element ${key} component`, () => {
    test('Has weight and defaultConfig static attributes defined', () => {
      expect(Component.weight).toBeDefined()
      expect(Component.defaultConfig).toBeDefined()
    })

    test('Can be rendered without errors', () => {
      const component = create(<Component config={ Component.defaultConfig }/>)
    })

    test('Has a valid <name> attribute', () => { //what is valid? "q_{id}"
      if (!excludedKeys.includes(key)) { //these elements dont have name attribute
        const component = create(<Component config={ Component.defaultConfig } mode={ "not build" }/>)
        if (key === "Name") {//Name has 2 values; first name and last name. Test should reconsidered for multiple value fields
          const myTestObject = component.root
          const nameAttributes = myTestObject.findAllByProps({name:"q_0[firstName]"})
          const multipleNames = nameAttributes.concat(myTestObject.findAllByProps({name:"q_0[lastName]"}))
          expect(multipleNames.length).toBeGreaterThan(1)
        } else {
          const myTestObject = component.root
          //assuming defaultConfig have id:0
          const nameAttributes = myTestObject.findAllByProps({name:"q_0"})//if its static "q_0" it still passes. should be improved
          const nameCount = nameAttributes.length
          expect(nameCount).toBeGreaterThan(0)
        }
      }
    })
  })
}
