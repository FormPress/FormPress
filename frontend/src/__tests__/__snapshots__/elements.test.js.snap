// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches the snapshot 1`] = `
<button>
  Nothing to do for now
</button>
`;
